code/modules/projectiles/guns/energy/special.dm

```diff
@@ -317,6 +317,116 @@
+	can_holster = FALSE
+	slot_flags = SLOT_BACK
+	cell_type = /obj/item/stock_parts/cell/bsg
+	shaded_charge = 1
+	can_fit_in_turrets = FALSE //Crystal would shatter, or someone would try to put an empty gun in the frame.
+	var/obj/item/assembly/signaler/anomaly/flux/core = null
+	var/has_bluespace_crystal = FALSE
+	var/admin_model = FALSE //For the admin gun, prevents crystal shattering, so anyone can use it, and you dont need to carry backup crystals.
+
+/obj/item/gun/energy/bsg/Destroy()
+	QDEL_NULL(core)
+	return ..()
+
+/obj/item/gun/energy/bsg/examine(mob/user)
+	. = ..()
+	if(core && has_bluespace_crystal)
+		. += "[src] is fully operational"
+	else if(core)
+		. += "It has a flux anomaly core installed, but no bluespace crystal in it."
+	else if(has_bluespace_crystal)
+		. += "It has a bluespace crystal installed, but no flux anomaly core in it."
+	else
+		. += "It is missing a flux anomaly core and bluespace crystal to be operational."
+
+/obj/item/gun/energy/bsg/attackby(obj/item/O as obj, mob/user as mob, params)
+	if(istype(O, /obj/item/stack/ore/bluespace_crystal))
+		if(has_bluespace_crystal)
+			to_chat(user, "<span class='notice'>[src] already has a bluespace crystal in it.</span>")
+			return
+		var/obj/item/stack/S = O
+		if(!loc || !S || S.get_amount() < 1)
+			return
+		to_chat(user, "<span class='notice'>You load [O] into [src].</span>")
+		S.use(1)
+		has_bluespace_crystal = TRUE
+		if(core)
+			icon_state = "BSG_FINISHED"
+		else
+			icon_state = "BSG_CRYSTAL"
+		return
+	if(istype(O, /obj/item/assembly/signaler/anomaly/flux))
+		if(core)
+			to_chat(user, "<span class='notice'>[src] already has a [O]!</span>")
+		if(!user.drop_item())
+			to_chat(user, "<span class='warning'>[O] is stuck to your hand!</span>")
+			return
+		to_chat(user, "<span class='notice'>You insert [O] into [src], and [src] starts to warm up.</span>")
+		O.forceMove(src)
+		core = O
+		if(has_bluespace_crystal)
+			icon_state = "BSG_FINISHED"
+		else
+			icon_state = "BSG_CORE"
+		return
+	else
+		return ..()
+
+/obj/item/gun/energy/bsg/process_fire(atom/target as mob|obj|turf, mob/living/user as mob|obj, message = 1, params, zone_override, bonus_spread = 0)
+	if(!has_bluespace_crystal)
+		to_chat(usr, "<span class='warning'>The [src] has no bluespace crystal to power it!</span>")
+		return 0
+	if(!core)
+		to_chat(usr, "<span class='warning'>The [src] has no flux anomaly core to power it!</span>")
+		return 0
+	..()
+/obj/item/gun/energy/bsg/process_chamber()
+	if(prob(25))
+		shatter()
+	..()
+	update_icon()
+
+/obj/item/gun/energy/bsg/emp_act(severity)
+	..()
+	if(prob(75 / severity))
+		if(has_bluespace_crystal)
+			shatter()
+
+/obj/item/gun/energy/bsg/proc/shatter()
+	if(admin_model)
+		return
+	visible_message("<span class='warning'>[src]'s bluespace crystal shatters!</span>")
+	playsound(src, 'sound/effects/pylon_shatter.ogg', 50, TRUE)
+	has_bluespace_crystal = FALSE
+	if(core)
+		icon_state = "BSG_CORE"
+	else
+		icon_state = "BSG"
```

Comment:
I feel like you should override `update_icon()` and change the item's sprite according to the various states, so that you only do them in a singular place.